### DISCUSSION TITLE
Benchmarks for lossless transformations

### DISCUSSION THREAD

**Discussion by a-apiko** [libvips/libvips#3601] [Q&A]
I want to use libvips library to optimize some images without losing quality. I see [benchmarks](https://github.com/libvips/libvips/wiki/Benchmarks) for memory usage  but not for lossless transformations. Do you have some rough estimates of average memory saving? File extension should remain the same. Or maybe at least some link to huge asset of different images with different formats for benchmarking purposes? 
Thanks for your library and help

**Comment by jcupitt** [comment]
Hi @a-apiko, 

libvips isn't an image size optimisation library. File format optimisations are very specific to the format and require really detailed knowledge of all the quirks and tricks. You'd be much better off with an optimiser designed around that format, so pngcrush, the various gif crushers, the jpegtran commands, and so on.

Like PIL, numpy, imagemagick, and friends, libvips always unpacks to uncompressed RGB / G / CMYK / etc. on load, and always regenerates the output file from scratch on save. It has no support for directly transforming compressed images.

Having said that, there are some cases where the libvips savers can improve image compression -- it can usually produce a better GIF or PNG palette, and it supports mozjpeg, which can shrink JPEG files by 30% or so. You'd need to give some specific examples.

**Reply by a-apiko** [reply]
Hi @jcupitt,
Thanks for answer. I don't have specific examples, just in general we have a lot of uncompressed images of different formats and have idea to use libvips (or any other library) to compress them to save some storage, but before to calculate does it make any sense (is memory saving is significant). I saw lossless options while using libvips and saving in different formats and thought that it's common task for which libvips is used

**Reply by jcupitt** [reply]
Oh, you have uncompressed images you want to save in a lossless format? I misunderstood.

You should be able to get maybe a 30% size reduction, but access will be slower. Factors to consider:

- size reduction (test with a range of images, including photos and line art)
- do you need to support more than 8 bits per channel, or more than 3 channels, or large images
- compress time
- decompress time
- what software needs to consume these images? libvips can write some strange formats (eg. tiled tiff with lossless webp compression) and not all software will be able to read them

I would look into png (works everywhere, but mediocre compression), webp (reasonable support in most software, but size limits), and jxl (very good lossless compression, not widely supported, size limits). You could look at TIFF as well, but you need to be quite expert to get good results.

**Reply by jcupitt** [reply]
For example, with a large uncompressed photo:

```
$ vips copy nina.jpg x.v
$ time vips copy x.v x.jxl[lossless]

real	0m24.523s
user	0m39.644s
sys	0m2.140s
$ time vips copy x.v x.png

real	0m1.088s
user	0m1.064s
sys	0m0.080s
$ ls -l x.png x.jxl x.v
-rw-r--r-- 1 john john  5992774 Aug 11 14:27 x.jxl
-rw-r--r-- 1 john john 22456754 Aug 11 14:27 x.png
-rw-r--r-- 1 john john 73203045 Aug 11 14:26 x.v
```

PNG compresses in 1s and compresses by more than a factor of 3. JXL takes 25s to compress, but makes a file more than 10x smaller.

With line art, you'll probably find PNG and JXL are pretty similar in size reduction, but PNG will be a lot faster.

**Reply by a-apiko** [reply]
Thanks for detailed explanation!

**do you need to support more than 8 bits per channel, or more than 3 channels, or large images:**
I'm not sure, in general it's just users data, and main criteria is to make images smaller but transformation should be lossless if it's not possible for specific image then I think we don't touch it. Isn't it can be somehow auto-detected? lossless parameter wouldn't be enough for these cases?

**what software needs to consume these images? libvips can write some strange formats (eg. tiled tiff with lossless webp compression) and not all software will be able to read them**
It's for browser usage in general, maybe we have some specific formats but we don't think about displaying them, it's users data and users know how to open them

 **PNG compresses in 1s and compresses by more than a factor of 3. JXL takes 25s to compress, but makes a file more than 10x smaller. With line art, you'll probably find PNG and JXL are pretty similar in size reduction, but PNG will be a lot faster.**
So in general it's just better to convert everything to png? Currently I think we need to produce the same extension of file as input, but convert them to png sounds also quite good
